Add basic deck submission workflow

levinuss opened the pull request:
where a deck can be submitted for review by an admin or moderator and then to be shown to all users in the create session view.

TODO / open questions:

- [ ] move check for `public-rw-listed` to the JS instead of the API or add specific API call for that filter (should probably not be frontend)
- [x] check for mobile layout

closes #735

schu commented:
After thinking about this for a bit: I would prefer if we have a separate model (and corresponding controller) `DeckSubmission` that references (one-to-one) the submitted deck and (also one-to-one) the submitting user. That feels cleaner and more elegant to me and it would allow us to identify the submitter, time of submission (through the default `created_at` column) and potentially extend submissions later. For example, we might want to keep accepted submissions for "user achievements".

levinuss commented:
@schu Also put some thought in it, this is probably the way intended by Lavarel for this use-case, also you are right, it feels better :)
I reworked it to have a `DeckSubmission` model and resource controller.
Maybe you can look at it and tell me about improvements if it is the right direction in general.


> and (also one-to-one) the submitting user

A user should probably be able to have multiple pending submissions, so I made this a `hasMany` relationship from the user side.

schu commented on resources/views/submissions.blade.php:
```diff
@@ -0,0 +1,82 @@
+
+<div class="row pt-4">
+    @forelse ($submissions as $submission)
+        <div class="col-md-3 mb-1">
```
Can we make this at least `col-md-6`? I think `col-md-3` doesn't look right and very crammed.

levinuss replied:
For me `col-md-6` looks very widespread with a lot of unused white space. Still changed it to `col-md-6` for review.
If you like it more like this we can leave it at `col-md6`.

schu commented on resources/views/submissions.blade.php:
```diff
@@ -0,0 +1,82 @@
+                </div>
+                <div class="card-body">
+                    <h6 class="card-title text-nowrap overflow-hidden" title="Name">
+                        <i class="bi bi-file-font"></i> {{ $submission->deck->name }}
```
I don't like the icon and would like to stick to the layout that we use on the other pages (new session, etc.), i.e. have only the name.

schu commented on resources/views/submissions.blade.php:
```diff
@@ -0,0 +1,82 @@
+                    <h6 class="card-title text-nowrap overflow-hidden" title="Name">
+                        <i class="bi bi-file-font"></i> {{ $submission->deck->name }}
+                    </h6>
+                    <h6 class="card-title text-nowrap overflow-hidden" title="Module">
```
I like having the meta data here but would prefer to not use `card-title` elements for those.

Something like this maybe, we just a `<hr><p> ... <br> ... </p><hr>`:

![image](https://github.com/openmultiplechoice/openmultiplechoice/assets/366468/7d42aab6-5a23-4d18-8cff-0136ade8c8ce)

I also like the file icon with ear :)

schu replied:
I'm not sure about the "visual language" regarding icons yet, but that's something that we can decide and change later (e.g. where should we use icons, which icons, etc).

levinuss replied:
Yep, using `card-title` is a copypasta mistake.
I like the visual break with `<hr>` and the button arrangement in your proposal.
So for the icon should we go with `file-earmark` or `file-earmark-text` then?

schu replied:
`file-earmark-check-fill` maybe?

levinuss replied:
Would you also be fine with `file-earmark-check` (without fill)? I think it aligns better with the other icons as they are not filled as well.

schu replied:
You decide :) We'll review the overall design at a later point and could change it to something else then.

schu commented on resources/views/submissions.blade.php:
```diff
@@ -0,0 +1,82 @@
+                                @csrf
+
+                                <button class="btn btn-sm btn-primary form-control" type="submit">
+                                    <i class="bi bi-file-arrow-up-fill"></i> Approve
```
Regarding the button placement, so far we always have the "positive action" on the left side, if I'm not mistaken. So I would prefer the placement as shown in the mockup above (Approve, Browse, Delete).

schu commented on routes/web.php:
```diff
@@ -1,5 +1,7 @@
 <?php
 
+use App\Http\Controllers\DeckSubmissionController;
+use App\Models\Deck;
```
As mentioned above, let's stick to the convention regarding use statements, i.e. those should be in their designated blocks below.

schu commented on resources/views/deck-editor.blade.php:
```diff
@@ -41,6 +41,26 @@
 
 @if ($deck->user_id == Auth::id() || Auth::user()->is_admin)
```
If the deck is `public-rw-listed` already, we should probably hide both the deck submission and deck archive areas (i.e. the "Danger Zone"). The deck cannot be re-submitted and users shouldn't be able to archive  `public-rw-listed` decks.

levinuss replied:
For archiving, we probably also need a backend check for `$deck->access != 'public-rw-listed'` (which should not be part of this PR?)

schu replied:
I would prefer it as part of this PR:

* Hide submission and archiving in the deck form if it's a `public-rw-listed` deck and
* Add the check to the DeckController

levinuss replied:
Had already done the frontend check (which hides the danger zone for now) and also added a check for `is_archived` to `Controllers/DeckController.php`. Should we also add it to the API Controller?

schu replied:
In general, the API controller should show the same behavior and do the check, yes (unless there's a good reason to handle an API request differently here?).

levinuss replied:
I don't think so. You should not be able to archive a listed deck.
I copied the code for the check to the API controller (and adjusted error handling) but don't really know how to test it tbh

schu commented on app/Http/Controllers/DeckSubmissionController.php:
```diff
@@ -0,0 +1,90 @@
+            abort(403, 'Unauthorized');
+        }
+        $submissions = DeckSubmission::all();
+        return view('submissions', ['submissions' => $submissions]);
```
Nitpick: please name the Blade template `deck-submissions` instead of `submissions` to go along with the naming for other deck templates.

schu commented on resources/views/deck.blade.php:
```diff
@@ -10,7 +10,7 @@
 <div class="row mb-3">
     <div class="col">
-        <h1 class="h4">{{ $deck->name }}</h1>
+        <h1 class="h4">{{ $deck->name }}@if($deck->submission) <span class="badge text-rounded-pill text-bg-info">Submitted for public listing</span>@endif</h1>
```
I would prefer a shorter tag and to use `secondary` instead of `info`, i.e.

```
<h1 class="h4">
    {{ $deck->name }}
    @if($deck->submission)
        <span class="badge text-rounded-pill text-bg-secondary">Submitted</span>
    @endif
</h1>
```

Something to be discussed outside of this PR: where should be put and show the label? I'm also not sure yet if the tag should be so bold.

schu reviewed (changes requested):
Thanks! Looks good to me overall, comments on mostly small things inline.

schu commented:
> A user should probably be able to have multiple pending submissions, so I made this a `hasMany` relationship from the user side.

Correct, wrong thinking on my side :)

schu commented:
> have checks (e.g. whether a module is selected) before submission

Right now I don't think this is necessary, otherwise we can add checks later.

schu commented:
> check for mobile layout

In Firefox / Chrome you can check the responsive design with [Responsive Design Mode](https://firefox-source-docs.mozilla.org/devtools-user/responsive_design_mode/index.html). In my experience, that's often "good enough" and device / mobile specific quirks not very common.

schu commented:
I think it would be helpful to show an indicator in the menu (both for the top-level `Admin` and `Submissions` section) if there are pending submissions, like the red dot shown for the "Profile" example button here: https://getbootstrap.com/docs/5.3/components/badge/#positioned Maybe not `bg-danger` but only `bg-secondary` (to "save" `bg-danger` for more important things).

Can also be part of a second PR, as you prefer.

levinuss commented:
> With Firefox / Chrome you can check the responsive design with [Responsive Design Mode](https://firefox-source-docs.mozilla.org/devtools-user/responsive_design_mode/index.html) on the desktop. In my experience, that's often "good enough" and device / mobile specific quirks not very common.

Well, the button arrangement does not work that well on mobile. Do you have any suggestions on how to enhance this or should we leave it like this? I just added some vertical padding for now.



> I think it would be helpful to show an indicator in the menu (both for the top-level `Admin` and `Submissions` section) if there are pending submissions, like the red dot shown for the "Profile" example button here: https://getbootstrap.com/docs/5.3/components/badge/#positioned Maybe not `bg-danger` but only `bg-secondary` (to "save" `bg-danger` for more important things).

I added it to this PR. I do not like the pill in the upper right corner in the menu because it touches the site border. Should we use the usual pill indicator (in-line with `Admin`) for this?

schu commented:
>> I think it would be helpful to show an indicator in the menu

> I added it to this PR. I do not like the pill in the upper right corner in the menu because it touches the site border.

Indeed, the indicator looks odd and misplaced, let's not add it and think about notification indicators later outside of this PR.

schu commented:
> Well, the button arrangement does not work that well on mobile. Do you have any suggestions on how to enhance this or should we leave it like this? I just added some vertical padding for now.

You mean the fact that they get stacked? I think that's OK for the moment.

![image](https://github.com/openmultiplechoice/openmultiplechoice/assets/366468/a096229a-86a3-4f6b-9dbd-4b30677eb9c5)

schu reviewed (changes requested):
As commented below, let's not do the menu indicator for now and add the checks to remove submission/archiving for `public-rw-listed` decks, otherwise good to merge, I think.

schu commented on app/Http/Controllers/DeckController.php:
```diff
@@ -37,6 +37,8 @@ public function show(Deck $deck)
         $questions = $deck->questions()->orderBy('id', 'asc')->get();
         $nextQuestion = $questions->first();
+        // Load the submission relationship to check if the deck has been submitted (used in template)
+        $deck = $deck->load('submission');
```
```
$deck->load('submission');
```

is enough, no need to redefine the variable. Same below.

schu commented on app/Http/Controllers/DeckSubmissionController.php:
```diff
@@ -0,0 +1,86 @@
+            abort(403, 'Unauthorized');
+        }
+        // If the deck was already submitted or is already listed publicly, return an error message.
+        if (DeckSubmission::where('deck_id', $request->deck_id)->exists() or
```
```
DeckSubmission::where('deck_id', $request->deck_id)
```

isn't necessary, you could simply write `$deck->submission` to the same effect (an extra DB query), the `submission` property would be [lazy loaded](https://laravel.com/docs/10.x/eloquent-relationships#relationship-methods-vs-dynamic-properties) then.

But to avoid an extra DB request, let's eager load the submission above:

```
Deck::with('submission')->findOrFail(...)
```

and then use `$deck->submission->exists()`.

schu commented on resources/views/deck-editor.blade.php:
```diff
@@ -39,8 +39,28 @@
 </div>
 
-@if ($deck->user_id == Auth::id() || Auth::user()->is_admin)
+@if ($deck->user_id == Auth::id() || Auth::user()->is_admin && $deck->access != 'public-rw-listed')
```
If the original owner opens the page, `$deck->user_id == Auth::id()` evaluates to true and thus the if statement is true, i.e. we show the forms where we don't want to.

I think it's clearer if we start with the condition that always should be true:

```
if ($deck->access != 'public-rw-listed' && ($deck->user_id == Auth::id() || Auth::user()->is_admin)) ...
```

levinuss replied:
Thanks for clarifying, I had some issues with finding out how this logic works in PHP and using parentheses for nested conditions didn't seem to work for me (but it does...).

levinuss commented on app/Http/Controllers/DeckController.php:
```diff
@@ -68,6 +72,12 @@ public function update(Request $request, Deck $deck)
         }
 
+        // You should not be able to archive the deck if it was submitted or listed
+        $is_submitted_or_listed = ($deck->submission->exists() || $deck->access == 'public-rw-listed');
```
> i.e. we also need to load the submission before ($deck->load('submission');).

That alone does not solve the problem for me as `$deck->submission` is still `null` if there is no submission, even if we load the submission before. It works now for me after I added the `load` and removed `exists()`
